File: day4/d4.py
def read_word_search(file_path):
    with open(file_path, 'r') as file:
        return [line.strip() for line in file.readlines()]

def count_word_occurrences(grid, word):
    rows, cols = len(grid), len(grid[0])
    word_len = len(word)
    count = 0

    def check_direction(x, y, dx, dy):
        for i in range(word_len):
            if not (0 <= x + i * dx < rows and 0 <= y + i * dy < cols):
                return False
            if grid[x + i * dx][y + i * dy] != word[i]:
                return False
        return True

    for i in range(rows):
        for j in range(cols):
            if grid[i][j] == word[0]:
                directions = [(1, 0), (0, 1), (1, 1), (1, -1), (-1, 0), (0, -1), (-1, -1), (-1, 1)]
                for dx, dy in directions:
                    if check_direction(i, j, dx, dy):
                        count += 1

    return count

if __name__ == "__main__":
    file_path = 'day4Input.txt'
    word_search = read_word_search(file_path)
    word = "XMAS"
    occurrences = count_word_occurrences(word_search, word)
    print(f"The word '{word}' appears {occurrences} times in the word search.")

File: day4/d4.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"
)

func readWordSearch(filePath string) ([]string, error) {
    file, err := os.Open(filePath)
    if err != nil {
        return nil, err
    }
    defer file.Close()

    var grid []string
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        grid = append(grid, strings.TrimSpace(scanner.Text()))
    }

    if err := scanner.Err(); err != nil {
        return nil, err
    }

    return grid, nil
}

func countWordOccurrences(grid []string, word string) int {
    rows, cols := len(grid), len(grid[0])
    wordLen := len(word)
    count := 0

    checkDirection := func(x, y, dx, dy int) bool {
        for i := 0; i < wordLen; i++ {
            nx, ny := x+i*dx, y+i*dy
            if nx < 0 || nx >= rows || ny < 0 || ny >= cols || grid[nx][ny] != word[i] {
                return false
            }
        }
        return true
    }

    directions := [][2]int{{1, 0}, {0, 1}, {1, 1}, {1, -1}, {-1, 0}, {0, -1}, {-1, -1}, {-1, 1}}

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            if grid[i][j] == word[0] {
                for _, dir := range directions {
                    if checkDirection(i, j, dir[0], dir[1]) {
                        count++
                    }
                }
            }
        }
    }

    return count
}

func main() {
    grid, err := readWordSearch("day4Input.txt")
    if err != nil {
        fmt.Println("Error reading file:", err)
        return
    }

    word := "XMAS"
    occurrences := countWordOccurrences(grid, word)
    fmt.Printf("Number of occurrences of the word \"%s\": %d\n", word, occurrences)
}